src/torch/README.md

```diff
@@ -0,0 +1,123 @@
+
+*   **Reverse Mode Autodiff:**  In reverse mode, the gradients are computed by traversing the computational graph backwards, from the output (typically a scalar loss function) to the inputs.  This is highly efficient for functions with many inputs and a single output, which is common in machine learning.
+
+*   **Jacobian-Vector Product (JVP):**  Let `y = f(x)` be a vector-valued function, where `x` is an input vector and `y` is an output vector. The Jacobian matrix `J` contains all the first-order partial derivatives of `f`.  The JVP is the product `Jv`, where `v` is a vector.  It gives the directional derivative of `f` at `x` in the direction of `v`.  *Forward mode* autodiff computes JVPs.
+
+*   **Vector-Jacobian Product (VJP):** With `y = f(x)` as above, the VJP is the product `w^T J`, where `w` is a vector (often called the "upstream/puput gradient").  The VJP represents the gradient of a scalar-valued function that depends on `y` (e.g., a loss function), with respect to the input `x`.  *Reverse mode* autodiff computes VJPs. 
+
+*   **Hessian-Vector Product (HVP):** The Hessian matrix `H` contains all the second-order partial derivatives of a scalar-valued function `f(x)`. The HVP is the product `Hv`, where `v` is a vector.  It gives the second-order directional derivative of `f` at `x` in the direction of `v`. 
+
+## Wrapper Design
+
+The wrapper provides three main classes:
+
+1.  `SIRFTorchOperator`: Wraps a SIRF `Operator` (e.g., a projection operator). Computes the JVP forward and VJP in reverse.
+2.  `SIRFTorchObjectiveFunction`: Wraps a SIRF `ObjectiveFunction` for calculating its value forward, and gradient in reverse mode.
+3.  `SIRFTorchObjectiveFunctionGradient`: Wraps a SIRF `ObjectiveFunction` for calculating its gradient forward and computing the Hessian-vector product in reverse mode.
+
+These classes use custom `torch.autograd.Function` implementations (`_Operator`, `_ObjectiveFunction`, and `_ObjectiveFunctionGradient`) to define the forward and backward passes, handling the conversions between PyTorch tensors and SIRF objects.
+
+### `_Operator` (Forward and Backward Passes)
+
+*   **Forward Pass:**
+    1.  Converts the input PyTorch tensor to a SIRF object.
+    2.  Applies the SIRF `Operator.forward()` method.  Let's represent the SIRF Operator as `A`.  So, this step computes `y = Ax`, where `x` is the input (SIRF object) and `y` is the output (SIRF object).
+    3.  Converts the result back to a PyTorch tensor.
+    4.  If the input tensor requires gradients, it saves relevant information (the output SIRF object and the operator) in the context (`ctx`) for use in the backward pass.
+
+*   **Backward Pass (Adjoint Operation - VJP):**
+    1.  Receives the "upstream gradient" (`grad_output`) – the gradient of the loss `L` with respect to the *output* of the forward pass: `∂L/∂y`.
```

Comment:
I find the `grad_output` name very confusing, but maybe it's standard in torch doc. In any case, I would remove any reference to loss 'L'. Who knows where this is used. It could easily be part of a chain of operations for instance. If you want to keep it, then I would say "for instance, if this is part of a sequence of operations like `L(operator(x))`... (but does that really help?)